experimental/parser/parse.go

```diff
@@ -0,0 +1,70 @@
+// Copyright 2020-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package parser
+
+import (
+	"github.com/bufbuild/protocompile/experimental/ast"
+	"github.com/bufbuild/protocompile/experimental/internal/taxa"
+	"github.com/bufbuild/protocompile/experimental/report"
+	"github.com/bufbuild/protocompile/experimental/token"
+)
+
+// Parse lexes and parses the Protobuf file tracked by ctx.
+//
+// Diagnostics generated by this process are written to errs. Returns whether
+// parsing succeeded without errors.
+//
+// Parse will freeze the stream in ctx when it is done.
+func Parse(ctx ast.Context, errs *report.Report) (file ast.File, ok bool) {
+	prior := len(errs.Diagnostics)
+
+	Lex(ctx, errs)
```

Comment:
Sure, we can do that. If it turns out that's not the right modality we can change this later.

Reply:
Maybe a TODO? I see that you unexported `lex` and changed the input argument to a `*report.File`. But I still think the `lex` function should take the file and return a `token.Context` and that `parse` should accept a `token.Context` and return an `ast.Context`. Right now the token and ast context's are one type (or rather `ast.Context` implements `token.Context`, but they seem trivially separable and that would definitely make these easier to read -- functions taking inputs and producing results instead of taking a bag of data that gets mutated.